src/ray/raylet/main.cc

```diff
@@ -224,6 +225,47 @@ int main(int argc, char *argv[]) {
 #endif
   };
 
+  auto shutted_down = std::make_shared<std::atomic<bool>>(false);
+
+  auto shutdown_raylet_after_unregistration =
+      [&main_service, &raylet_socket_name, &raylet, &gcs_client]() {
+        // We should stop the service and remove the local socket file.
+        raylet->Stop();
+        gcs_client->Disconnect();
+        ray::stats::Shutdown();
+        main_service.stop();
+        remove(raylet_socket_name.c_str());
+      };
+
+  // Shut down raylet gracefully, in a synchronous fashion.
+  // This is an internal method and should only be run on the main_service.
+  auto _shutdown_raylet_gracefully_sync =
```

Comment:
```suggestion
  auto shutdown_raylet_gracefully_internal =
```

It's not really sync since UnregisterSelf is async?